Mathlib/RingTheory/PiTensorProduct.lean

```diff
@@ -0,0 +1,295 @@
+/-
+Copyright (c) 2024 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import Mathlib.LinearAlgebra.PiTensorProduct
+import Mathlib.Algebra.Algebra.Hom
+import Mathlib.Algebra.Group.Pi
+
+/-!
+# Tensor product of `R`-algebras and rings
+
+If `(Aᵢ)` is a family of `R`-algebras then the `R`-tensor `⨂ᵢ Aᵢ` is an `R`-algebra as well with
+its structure map defined by `r ↦ r • 1`.
+
+In particular if we take `R` to be `ℤ`, then this collapse into tensor product of rings.
+-/
+
+open TensorProduct Function
+
+variable {ι R : Type*} {A : ι → Type*}
+
+namespace PiTensorProduct
+
+noncomputable section AddCommMonoidWithOne
+
+variable [CommSemiring R] [∀ i, AddCommMonoidWithOne (A i)] [∀ i, Module R (A i)]
+
+instance one : One (⨂[R] i, A i) where
+  one := tprod R 1
+
+instance addCommMonoidWithOne : AddCommMonoidWithOne (⨂[R] i, A i) where
+  __ := inferInstanceAs (AddCommMonoid (⨂[R] i, A i))
+  __ := one
+
+end AddCommMonoidWithOne
+
+noncomputable section NonUnitalNonAssocSemiring
+
+variable [CommSemiring R] [∀ i, NonUnitalNonAssocSemiring (A i)]
+variable [∀ i, Module R (A i)] [∀ i, SMulCommClass R (A i) (A i)] [∀ i, IsScalarTower R (A i) (A i)]
+
+/--
+The multiplication in tensor product of rings is induced by `(xᵢ) * (yᵢ) = (xᵢ * yᵢ)`
+-/
+def lmul : (⨂[R] i, A i) →ₗ[R] (⨂[R] i, A i) →ₗ[R] (⨂[R] i, A i) :=
+lift
+{ toFun := fun x ↦ lift
+    { toFun := fun y ↦ tprod _ (x * y)
+      map_add' := fun z i a a' ↦ by
```

Comment:
I think there's some missing infrastructure here; can you first define `PiTensorProduct.map` to match `TensorProduct.map`? Then I think you might be able to take some ideas from #6195.